pam/gdm/protocol.go

```diff
@@ -0,0 +1,170 @@
+// Package gdm is the package for the GDM pam module handing.
+package gdm
+
+//TODO: Watch https://github.com/protocolbuffers/protobuf for any changes on the experimental status of optional fields,
+// previously described on: https://github.com/protocolbuffers/protobuf/blob/main/docs/implementing_proto3_presence.md.
+//
+// Should it become default, remove the --experimental_allow_proto3_optional flag from the go generate command below.
+//go:generate sh -c "PATH=\"$PATH:`go env GOPATH`/bin\" protoc --proto_path=. --go_out=. --go_opt=paths=source_relative gdm.proto --experimental_allow_proto3_optional -I../.."
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"slices"
+
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	// ProtoVersion is the version of the JSON protocol.
+	ProtoVersion = uint32(1)
+)
+
+// Request is an interface implementing all the gdm requests.
+type Request = isRequestData_Data
+
+// Response is an interface implementing all the gdm responses.
+type Response = isResponseData_Data
+
+// Event is an interface implementing all the gdm events.
+type Event = isEventData_Data
+
+// NewDataFromJSON unmarshals data from json bytes.
+func NewDataFromJSON(bytes []byte) (*Data, error) {
+	var gdmData Data
+	if err := protojson.Unmarshal(bytes, &gdmData); err != nil {
+		return nil, err
+	}
+
+	if err := gdmData.Check(); err != nil {
+		return nil, err
+	}
+
+	return &gdmData, nil
+}
+
+// FIXME: Do not do this when building the module in release mode, this is
+// just relevant for testing purposes.
+func (d *Data) checkMembers(acceptedMembers []string) error {
+	//nolint:govet
+	//We only redirect the value to figure out its type.
+	val := reflect.ValueOf(*d)
+	typ := val.Type()
+	acceptedMembers = append(acceptedMembers, []string{
+		"Type", "state", "sizeCache", "unknownFields",
+	}...)
+
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		fieldType := typ.Field(i)
+
+		if slices.Contains(acceptedMembers, fieldType.Name) {
+			continue
+		}
+
+		if !field.IsZero() {
+			return fmt.Errorf("field %v should not be defined", fieldType.Name)
+		}
+	}
+
+	return nil
+}
+
+// Check allows to check the sanity of a data value.
+func (d *Data) Check() error {
+	switch d.Type {
+	case DataType_unknownType:
+		return fmt.Errorf("unexpected type %v", d.Type.String())
+
+	case DataType_hello:
+		if err := d.checkMembers([]string{"Hello"}); err != nil {
+			return err
+		}
```

Comment:
Following the logic of the previous function, this shouldn't be added to the release binary as well (?)

Reply:
Mh, yeah that's possible, however calling a dummy function is probably not a big deal comparing to duplicating some code, but I can see how to manage that cleanly.